### PR TITLE
Update installation instructions

### LINKED ISSUE #904: Update installation doc to include details about Docker deployment
Existing documentation: https://elyra.readthedocs.io/en/latest/getting_started/installation.html

See JupyterLab doc for reference https://jupyterlab.readthedocs.io/en/stable/getting_started/installation.html

### PR DESCRIPTION
**Pull request by ptitzler**
This PR restructures the content of the installation documentation and adds details on how to run Elyra on Docker. The layout mimics the structure of the [JupyterLab installation guide](https://jupyterlab.readthedocs.io/en/stable/getting_started/installation.html).

image

Closes #904 
 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

File: docs/source/getting_started/installation.md [hunk 1]
```diff
@@ -16,43 +16,52 @@ limitations under the License.
 {% endcomment %}
 -->
 ## Installation
-Elyra can be installed via PyPI:
+
+The Elyra [JupyterLab extension](https://jupyterlab.readthedocs.io/en/stable/user/extensions.html) can be installed using `pip`, `conda`, or from source code. Installation of Elyra requires a rebuild of JupyterLab.
```
Comment (lresende) [inline docs/source/getting_started/installation.md:21]: I would say something like, `JupyterLab currently requires a build after installing any extension.`

File: docs/source/getting_started/installation.md [hunk 2]
```diff
@@ -87,3 +102,33 @@ Known labextensions:
         @jupyterlab/toc v4.0.0  enabled  OK
         nbdime-jupyterlab v2.0.0  enabled  OK
 ```
+
+### Docker 
+
+If you have Docker installed, you can install and use JupyterLab and Elyra by selecting one of the [ready-to-run Docker images](https://hub.docker.com/r/elyra/elyra/tags) maintained by the Elyra Team:
```
Comment (lresende) [inline docs/source/getting_started/installation.md:108]:
```suggestion
If you have Docker installed, you can use JupyterLab and Elyra by running one of the [ready-to-run Docker images](https://hub.docker.com/r/elyra/elyra/tags) maintained by the Elyra Team:
```

File: docs/source/getting_started/installation.md [hunk 3]
```diff
@@ -87,3 +102,33 @@ Known labextensions:
...
+```
+docker run -it -p 8888:8888 elyra/elyra:dev jupyter lab --debug
+```
+
+To make a local directory containing your Notebooks (e.g. `${HOME}/opensource/jupyter-notebooks/`) available in your Docker container, you can use a mount command similar to the following:
+
+```
+docker run -it -p 8888:8888 -v ${HOME}/opensource/jupyter-notebooks/:/home/jovyan/work -w /home/jovyan/work elyra/elyra:dev jupyter lab --debug
+```
+
```
Comment (lresende) [inline docs/source/getting_started/installation.md:125]: How about adding a new section to mount the available metadata from local system to docker? Just follow the same pattern above and point to `jupyter --data-dir/metadata`

File: docs/source/getting_started/installation.md (hunk 1 above)
Comment (ptitzler) [reply docs/source/getting_started/installation.md:21]: Changed

File: docs/source/getting_started/installation.md (hunk 3 above)
Comment (ptitzler) [reply docs/source/getting_started/installation.md:125]: As suggested this doesn't work unless I am missing something:

```
docker run -it -p 8888:8888 -v $HOME/jupyter-notebooks/:/home/jovyan/work -w /home/jovyan/work -v $HOME/jupyter-metadata:/home/jovyan/.local/share/jupyter/metadata elyra/elyra:dev jupyter lab --debug
```

yields

```
[D 16:23:07.010 LabApp] Searching ['/home/jovyan/work', '/home/jovyan/.jupyter', '/opt/conda/etc/jupyter', '/usr/local/etc/jupyter', '/etc/jupyter'] for config files
[D 16:23:07.011 LabApp] Looking for jupyter_config in /etc/jupyter
[D 16:23:07.011 LabApp] Looking for jupyter_config in /usr/local/etc/jupyter
[D 16:23:07.011 LabApp] Looking for jupyter_config in /opt/conda/etc/jupyter
[D 16:23:07.011 LabApp] Looking for jupyter_config in /home/jovyan/.jupyter
[D 16:23:07.011 LabApp] Looking for jupyter_config in /home/jovyan/work
[D 16:23:07.013 LabApp] Looking for jupyter_notebook_config in /etc/jupyter
[D 16:23:07.014 LabApp] Loaded config file: /etc/jupyter/jupyter_notebook_config.py
[D 16:23:07.014 LabApp] Looking for jupyter_notebook_config in /usr/local/etc/jupyter
[D 16:23:07.014 LabApp] Looking for jupyter_notebook_config in /opt/conda/etc/jupyter
[D 16:23:07.015 LabApp] Loaded config file: /opt/conda/etc/jupyter/jupyter_notebook_config.json
[D 16:23:07.015 LabApp] Looking for jupyter_notebook_config in /home/jovyan/.jupyter
[D 16:23:07.015 LabApp] Loaded config file: /home/jovyan/.jupyter/jupyter_notebook_config.py
[D 16:23:07.015 LabApp] Looking for jupyter_notebook_config in /home/jovyan/work
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/traitlets/traitlets.py", line 528, in get
    value = obj._trait_values[self.name]
KeyError: 'runtime_dir'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/conda/bin/jupyter-lab", line 8, in <module>
    sys.exit(main())
  File "/opt/conda/lib/python3.7/site-packages/jupyter_core/application.py", line 268, in launch_instance
    return super(JupyterApp, cls).launch_instance(argv=argv, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/traitlets/config/application.py", line 663, in launch_instance
    app.initialize(argv)
  File "</opt/conda/lib/python3.7/site-packages/decorator.py:decorator-gen-7>", line 2, in initialize
  File "/opt/conda/lib/python3.7/site-packages/traitlets/config/application.py", line 87, in catch_config_error
    return method(app, *args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/notebook/notebookapp.py", line 1766, in initialize
    self.init_configurables()
  File "/opt/conda/lib/python3.7/site-packages/notebook/notebookapp.py", line 1380, in init_configurables
    connection_dir=self.runtime_dir,
  File "/opt/conda/lib/python3.7/site-packages/traitlets/traitlets.py", line 556, in __get__
    return self.get(obj, cls)
  File "/opt/conda/lib/python3.7/site-packages/traitlets/traitlets.py", line 535, in get
    value = self._validate(obj, dynamic_default())
  File "/opt/conda/lib/python3.7/site-packages/jupyter_core/application.py", line 99, in _runtime_dir_default
    ensure_dir_exists(rd, mode=0o700)
  File "/opt/conda/lib/python3.7/site-packages/jupyter_core/utils/__init__.py", line 13, in ensure_dir_exists
    os.makedirs(path, mode=mode)
  File "/opt/conda/lib/python3.7/os.py", line 221, in makedirs
    mkdir(name, mode)
PermissionError: [Errno 13] Permission denied: '/home/jovyan/.local/share/jupyter/runtime'
```

I can only get it to work if the mount point is `jupyterlab --data-dir` (`/home/jovyan/.local/share/jupyter` in the image):

```
docker run -it -p 8888:8888 -v $HOME/jupyter-notebooks/:/home/jovyan/work -w /home/jovyan/work -v $HOME/jupyter-metadata:/home/jovyan/.local/share/jupyter elyra/elyra:dev jupyter lab --debug
```

Comment (ptitzler) [reply docs/source/getting_started/installation.md:140]: I updated the documentation to mount the JupyterLab data directory.

Comment (lresende) [reply docs/source/getting_started/installation.md:140]: you are actually right, as the current image does not have the `jupyter data-dir` created... as you mentioned I could make the following to work:
```
docker run -it -p 8888:8888 -v $HOME/jupyter-notebooks/:/home/jovyan/work -w /home/jovyan/work -v $HOME/Library/Jupyter:/home/jovyan/.local/share/jupyter elyra/elyra:1.1.0 jupyter lab --debug

```

File: docs/source/getting_started/installation.md [hunk 4]
```diff
@@ -87,3 +104,46 @@ Known labextensions:
...
+Invocation example 2: Run the most recent Elyra development build in a Docker container and mount the existing local `$HOME/jupyter-notebooks/` directory as JupyterLab work directory. This enables you to make existing notebooks and other files available in the Docker container. Only files in this working directory are retained when the Docker container is stopped. 
+
+```
+docker run -it -p 8888:8888\
+ -v ${HOME}/jupyter-notebooks/:/home/jovyan/work\
+ -w /home/jovyan/work\
+ elyra/elyra:dev jupyter lab --debug
+```
+
+Invocation example 3: Same as above. In addition the existing local `$HOME/jupyter-data-dir` directory is mounted as the Jupyter data directory, preserving all user-defined Elyra metadata artifacts you create, such as code snippets, runtime configurations, or runtime images.
```
Comment (lresende) [inline docs/source/getting_started/installation.md:131]: Just to make sure, there is a command `jupyter --data-dir` that gives you the location of the data directory (e.g. `/Users/lresende/Library/Jupyter` in my macos dev environment`

Comment (ptitzler) [reply docs/source/getting_started/installation.md]: It turns out we were thinking about two different scenarios and my arbitrary directory name choice was ambiguous. For the purpose of this documentation `jupyter-data-dir` is a placeholder for any valid local directory name.
 - If `$HOME/jupyter-data-dir`  refers to an empty local directory JupyterLab in the Docker container is started in a "clean" environment. 
 - If `$HOME/jupyter-data-dir`  refers to the data directory of a local JupyterLab installation (`jupyter --data-dir`) then the Jupyter metadata is shared between the local instance and the instance in the Docker container. This is an advanced usage scenario.

I updated the document accordingly, which should hopefully address the concern.

**Review by lresende — APPROVED** [review]
LGTM

**Pull request merged**